chore(deps): bumping cdk from 1.149.0 to 1.160.0

jericht opened the pull request:
### Changes
- Bump CDK to 1.160.0
- Removes Node12 (EOL) from CI and adds Node18
- Locked `aws-sdk-mock` to 5.5.0 due to:
    - Changes in [v5.6.1](https://github.com/dwyl/aws-sdk-mock/commit/32f3179494ecf861f1189e2edc84204ab1abdd3e) that causes typescript errors:
        ```
        aws-rfdk: ../../node_modules/aws-sdk-mock/index.d.ts:60:90 - error TS1110: Type expected.
        aws-rfdk: 60 export type NestedClientFullName<C extends ClientName, NC extends NestedClientName<C>> = `${C}.${NC}`;
        ```
    - API change in 5.6.0 that breaks unit test `lib/lambdas/nodejs/asg-attach-eni/test/test.js`
        ```
        aws-rfdk: FAIL lib/lambdas/nodejs/asg-attach-eni/test/test.js
        aws-rfdk:   ● processes all correct records
        aws-rfdk:     expect(received).toStrictEqual(expected) // deep equality
        aws-rfdk:     Expected: {"DeviceIndex": 1, "InstanceId": "i-0000000000", "NetworkInterfaceId": "eni-000000000"}
        aws-rfdk:     Received: [Function callback]
        aws-rfdk:       111 |   expect(attachSpy).toHaveBeenCalledTimes(2);
        aws-rfdk:       112 |   expect(completeSpy).toHaveBeenCalledTimes(2);
        aws-rfdk:     > 113 |   expect(attachSpy.mock.calls[0][0]).toStrictEqual({
        aws-rfdk:           |                                      ^
        aws-rfdk:       114 |     DeviceIndex: 1,
        aws-rfdk:       115 |     InstanceId: 'i-0000000000',
        aws-rfdk:       116 |     NetworkInterfaceId: 'eni-000000000',
        aws-rfdk:       at Object.<anonymous> (lib/lambdas/nodejs/asg-attach-eni/test/test.ts:113:38)
        aws-rfdk:   ● abandons launch when attach fails
        aws-rfdk:     expect(received).toStrictEqual(expected) // deep equality
        aws-rfdk:     Expected: {"AutoScalingGroupName": "ASG-Name-1", "InstanceId": "i-0000000000", "LifecycleActionResult": "ABANDON", "LifecycleActionToken": "Action-Token-1", "LifecycleHookName": "Hook-Name-1"}
        aws-rfdk:     Received: [Function callback]
        aws-rfdk:       167 |
        aws-rfdk:       168 |   // THEN
        aws-rfdk:     > 169 |   expect(completeSpy.mock.calls[0][0]).toStrictEqual({
        aws-rfdk:           |                                        ^
        aws-rfdk:       170 |     AutoScalingGroupName: 'ASG-Name-1',
        aws-rfdk:       171 |     LifecycleHookName: 'Hook-Name-1',
        aws-rfdk:       172 |     InstanceId: 'i-0000000000',
        aws-rfdk:       at Object.<anonymous> (lib/lambdas/nodejs/asg-attach-eni/test/test.ts:169:40)
        ```
- Removed resolution for `node-gyp`:
    ```
    [jericht@dev-dsk-jericht-2a-f8c9ce54 aws-rfdk]$ yarn list node-gyp
    yarn list v1.22.17
    warning Filtering by arguments is deprecated. Please use the pattern option instead.
    └─ node-gyp@8.4.1
    Done in 0.75s.
    ```
- Put dependencies for packages in `tools` back into sync with CDK. The following package versions have been __downgraded__ to match the versions our build tools use:
    - `eslint`
    - `eslint-plugin-jest`
    - `@typescript-eslint/eslint-plugin`
    - `@typescript-eslint/parser`
    - `fs-extra`
- Updated unit test `endpoint.test.ts` that used the [CDK_NUMERIC_TOKEN example value in docs](https://docs.aws.amazon.com/cdk/v2/guide/tokens.html#tokens_number) to generate a token with `Lazy.number(...)` instead since the test was failing.
- In `tools`, added `any` type annotation to variables in `catch` blocks due to [breaking change in Typescript 4.4](https://github.com/microsoft/TypeScript/wiki/Breaking-Changes#using-unknown-in-catch-variables)
- Updates linter rule to verify the node engine is >= 14, and updates the node engines declared in our package.json files to >= 14

### Testing
- `yarn build`
- `yarn build && npx cdk synth` in each example
- `npx cdk deploy "*"` the basic example and verified it deployed successfully

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

ddneilson reviewed (dismissed):
LGTM! Thanks Jericho

jusiskin commented:
Should we also bump the Python example app dependencies?

jusiskin reviewed (changes requested):
Looks mostly good. Only things are:

1.  CDK 1.160.0 deprecated support for Node 12. We should
    1.  Fix up references in `README.md` and `CONTRIBUTING.md`
    2.  Surface the breaking change with conventional commits (can probably squash everything since that will happen anyways on merge)
2. Python example app CDK dependencies need to be bumped

jericht committed:
chore(deps): bumping cdk from 1.149.0 to 1.160.0

BREAKING CHANGE: The minimum supported version of Node.js has been
changed to Node.js 14

jericht commented:
@jusiskin Thanks for catching those. Updated the python CDK deps and the docs to mention minimum Node.js 14

jusiskin commented on tools/pkglint/lib/rules.ts:
```diff
@@ -376,11 +376,11 @@ export class NodeCompatibility extends ValidationRule {
-        message: `packages must support node version 12 and up, but ${atTypesNode} is declared`,
-        fix: () => pkg.addDevDependency('@types/node', '^12.18.3'),
+        message: `packages must support node version 14 and up, but ${atTypesNode} is declared`,
+        fix: () => pkg.addDevDependency('@types/node', '^14.14.31'),
```
should this be `14.15.0`?

jericht replied:
This is for the `@types/node` typing module and the latest for 14 is this one: https://www.npmjs.com/package/@types/node

jusiskin replied:
We discussed offline and this is okay. There are no newer type stubs for 14 and Node 14.15.0 didn't have any significant changes over 14.14.x ([ref](https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V14.md#14.15.0)).
